scripts/Stretch.py

```diff
@@ -0,0 +1,357 @@
+            vPosAfter = np.resize(np.roll(vPos,-i-1,0),(iextra,2))
+            good_triangle = True
+            dp = ((vPos[i]-vPosAfter)**2).sum(1)
+            if np.amax(dp) < d1*d1:
+                good_triangle = False
+            else:
+                i3 = np.argmax(dp>=d1*d1)
+            if good_triangle:
+                vPosBefore = np.resize(np.roll(vPos,-i,0)[::-1],(iextra,2))
+                dp = ((vPos[i]-vPosBefore)**2).sum(1)
+                if np.amax(dp) < d1*d1:
+                    good_triangle = False
+                else:
+                    i1 = np.argmax(dp>=d1*d1)
+            if good_triangle:
+                rd=((vPosAfter[i3]-vPosBefore[i1])**2).sum(0)
+                r=((vPos[i]-vPosBefore[i1])*(vPosAfter[i3]-vPosBefore[i1])).sum(0)
+                if np.fabs(r) < 1000.0*np.fabs(rd):
+                    r /= rd
+                else:
+                    r = 0.5
+                pp=vPosBefore[i1] + r*(vPosAfter[i3]-vPosBefore[i1])
+                dpp=np.sqrt(((pp-vPos[i])**2).sum(0))
+                if dpp > 0.001:
+                    vTrans[i] = vPos[i] - (d4/dpp)*(pp-vPos[i])
+
+            i = i + 1
+        return
+
+    def wideTurn(self,vPos,vTrans):
+        '''
+        We have to select three points in order to form a triangle
+        These three points should be far enough from each other to have
+        a reliable estimation of the orientation of the current turn
+        '''
+        d1 = self.line_width / 2.0
+        d4 = self.stretch
+        i1 = 0
+        i = 1
+        i3 = 2
+        while i+1<len(vPos):
+            good_triangle = True
+            dp = ((vPos[i]-vPos[i+1:])**2).sum(1)
+            if np.amax(dp) < d1*d1:
+                good_triangle = False
+            else:
+                i3 = i+1 + np.argmax(dp>=d1*d1)
+            if good_triangle:
+                dp = ((vPos[i]-vPos[i-1::-1])**2).sum(1)
+                if np.amax(dp) < d1*d1:
+                    good_triangle = False
+                else:
+                    i1 = i-1 - np.argmax(dp>=d1*d1)
+            if good_triangle:
+                rd=((vPos[i3]-vPos[i1])**2).sum(0)
+                r=((vPos[i]-vPos[i1])*(vPos[i3]-vPos[i1])).sum(0)
+                if np.fabs(r) < 1000.0*np.fabs(rd):
+                    r /= rd
+                else:
+                    r = 0.5
+                pp=vPos[i1] + r*(vPos[i3]-vPos[i1])
+                dpp=np.sqrt(((pp-vPos[i])**2).sum(0))
+                if dpp > 0.001:
+                    vTrans[i] = vPos[i] - (d4/dpp)*(pp-vPos[i])
+            i = i + 1
+        return
+
+    def pushWall(self,vPos,vTrans):
+        d2 = self.line_width / 2.0
+        d4 = self.stretch
+        mrot = np.array([[0,-1],[1,0]]) # Quarter turn
+        for i in range(len(vPos)):
+            i1 = i
+            i2 = i+1
+            if i2 == len(vPos):
+                i2 = i-1
+            xm = vPos[i1]
+            xperp = np.dot(mrot,vPos[i2]-vPos[i1])
```

Comment:
The objective of this computation is not to move the point, but to find for the current segment if material was already deposited at one side or at the other side of this segment. To test if material is present, I look at the position from the first point of the segment (it was initially the middle, but this doesn't have a big impact), in the perpendicular direction, at a distance of the line size. If one of the segments of already deposited material is near the test point, there is material at this side.

This is the slowest part of the algorithm. It takes few seconds in C++, much more in python.